# DATA PROCESSING
***

<p style="text-align: center;">
  <img src="../images/data-processing-methods.png" alt="Data Processing">
</p>
</p>


### Import Dependencies

In [3]:
import pandas as pd

***
##### Load Dataset

In [5]:
path = "../data/raw/ecommerce_data_analysis.xlsx"
dataset = pd.read_excel(path)

In [6]:
dataset.sample(4)

InvoiceNo StockCode                          Description  Quantity  \
104176    545152     22859             EASTER TIN BUNNY BOUQUET        12   
454747    575602    84849D                HOT BATHS SOAP HOLDER         2   
440757    574547     22576  SWALLOW WOODEN CHRISTMAS DECORATION         1   
253455    559163     21625             VINTAGE UNION JACK APRON         1   

               InvoiceDate  UnitPrice  CustomerID         Country  
104176 2011-02-28 12:47:00       1.65     15227.0  United Kingdom  
454747 2011-11-10 12:27:00       1.69     17059.0  United Kingdom  
440757 2011-11-04 15:08:00       0.85     17841.0  United Kingdom  
253455 2011-07-06 16:33:00       4.13         NaN  United Kingdom

In [7]:
dataset.shape

(541909, 8)

**For this dataset with the following columns:**

- `InvoiceNo` - (Unique identifier for transactions)
- `StockCode` - (Product code)
- `Description` - (Product name)
- `Quantity` - (Quantity purchased)
- `InvoiceDate` - (Date of transaction)
- `UnitPrice` - (Price per unit of the product)
- `CustomerID` - (Unique identifier for customers)
- `Country` - (Country of the customer)

In [10]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    541909 non-null  object        
 1   StockCode    541909 non-null  object        
 2   Description  540455 non-null  object        
 3   Quantity     541909 non-null  int64         
 4   InvoiceDate  541909 non-null  datetime64[ns]
 5   UnitPrice    541909 non-null  float64       
 6   CustomerID   406829 non-null  float64       
 7   Country      541909 non-null  object        
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 33.1+ MB


**Handling Missing Values**

In [12]:
# Checking initial shape of data
initial_shape = dataset.shape
print("Initial shape of dataset:", initial_shape)

missing_values = dataset.isnull().sum()
print("Missing values:\n", missing_values)

cleaned_dataset = dataset.dropna()

# Checking the cleaned dataset shape
cleaned_shape = cleaned_dataset.shape
print("Shape of cleaned dataset:", cleaned_shape)

dropped_rows = initial_shape[0] - cleaned_shape[0]
print("Number of rows dropped:", dropped_rows)

# checking if cleaned dataset still has any missing values
print("Missing values in cleaned dataset:\n", cleaned_dataset.isnull().sum())


Initial shape of dataset: (541909, 8)
Missing values:
 InvoiceNo           0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     135080
Country             0
dtype: int64
Shape of cleaned dataset: (406829, 8)
Number of rows dropped: 135080
Missing values in cleaned dataset:
 InvoiceNo      0
StockCode      0
Description    0
Quantity       0
InvoiceDate    0
UnitPrice      0
CustomerID     0
Country        0
dtype: int64


In [13]:
# Convert CustomerID to int
cleaned_dataset.loc[:, 'CustomerID'] = cleaned_dataset['CustomerID'].astype('Int64')

In [14]:
cleaned_dataset.head(3)

InvoiceNo StockCode                         Description  Quantity  \
0    536365    85123A  WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                 WHITE METAL LANTERN         6   
2    536365    84406B      CREAM CUPID HEARTS COAT HANGER         8   

          InvoiceDate  UnitPrice  CustomerID         Country  
0 2010-12-01 08:26:00       2.55     17850.0  United Kingdom  
1 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
2 2010-12-01 08:26:00       2.75     17850.0  United Kingdom

In [15]:
# Getting the count of unique customer IDs
unique_customer_count = cleaned_dataset['CustomerID'].nunique()

print("Number of Unique Customer IDs:", unique_customer_count)

Number of Unique Customer IDs: 4372


In [16]:
# Getting the count of unique stokecode
unique_customer_count = cleaned_dataset['StockCode'].nunique()

print("Number of StockCode:", unique_customer_count)

Number of StockCode: 3684


In [17]:
# Getting the count of unique InvoiceNo
unique_customer_count = cleaned_dataset['InvoiceNo'].nunique()

print("Number of InvoiceNo:", unique_customer_count)

Number of InvoiceNo: 22190


Dataset contains `406,829` rows and there are `22,190` unique InvoiceNo values, that would suggest that each invoice has, on average, about 18 items (406,829 / 22,190 ≈ 18.33). This would imply that each invoice corresponds to multiple line items or products.

In many e-commerce datasets, it's common for a single invoice to include multiple items, resulting in multiple rows for each invoice number. **If this is the case in your dataset, then having 22,190 unique InvoiceNo for 406,829 total rows seems valid.**

In [35]:
# Check for number of items per invoice
invoice_counts = cleaned_dataset.groupby('InvoiceNo').size()

print(invoice_counts.head(10))

InvoiceNo
536365     7
536366     2
536367    12
536368     4
536369     1
536370    20
536371     1
536372     2
536373    16
536374     1
dtype: int64


**Checking For Duplicate Rows**

In [38]:
duplicate_rows = cleaned_dataset.duplicated().sum()
print(f"Number of duplicate rows: {duplicate_rows}")

Number of duplicate rows: 5225


In [51]:
# Drop duplicates and assign back to the dataset variable
cleaned_dataset = cleaned_dataset.drop_duplicates()

print(f"Number of rows after removing duplicates: {cleaned_dataset.shape[0]}")

Number of rows after removing duplicates: 401604


**Checking For Negative Quantities**

In [53]:
negative_quantities = cleaned_dataset[cleaned_dataset['Quantity'] < 0]
print(negative_quantities)

       InvoiceNo StockCode                       Description  Quantity  \
141      C536379         D                          Discount        -1   
154      C536383    35004C   SET OF 3 COLOURED  FLYING DUCKS        -1   
235      C536391     22556    PLASTERS IN TIN CIRCUS PARADE        -12   
236      C536391     21984  PACK OF 12 PINK PAISLEY TISSUES        -24   
237      C536391     21983  PACK OF 12 BLUE PAISLEY TISSUES        -24   
...          ...       ...                               ...       ...   
540449   C581490     23144   ZINC T-LIGHT HOLDER STARS SMALL       -11   
541541   C581499         M                            Manual        -1   
541715   C581568     21258        VICTORIAN SEWING BOX LARGE        -5   
541716   C581569     84978  HANGING HEART JAR T-LIGHT HOLDER        -1   
541717   C581569     20979     36 PENCILS TUBE RED RETROSPOT        -5   

               InvoiceDate  UnitPrice  CustomerID         Country  
141    2010-12-01 09:41:00      27.50     1

The prefix "C" in some InvoiceNo values might indicate that these are credit notes or refunds. In an e-commerce context, negative quantities often indicate returns or cancellations of orders.

Checking if the `prefix` of the InvoiceNo for positive quantities differs from those with negative quantities

In [57]:
# Filtering positive quantities
positive_quantities = cleaned_dataset[cleaned_dataset['Quantity'] > 0].copy()  # Creating a copy to avoid SettingWithCopyWarning

# Ensure InvoiceNo is treated as a string
positive_quantities['InvoiceNo'] = positive_quantities['InvoiceNo'].astype(str)

# Extracting the prefix of InvoiceNo for positive quantities
positive_quantities['InvoicePrefix'] = positive_quantities['InvoiceNo'].str[0]

# Checking unique prefixes for positive quantities
positive_prefixes = positive_quantities['InvoicePrefix'].unique()
print("Unique prefixes for positive quantities:", positive_prefixes)

# Comparing with negative quantities (make sure negative_quantities is defined and has the same checks)
negative_prefixes = negative_quantities['InvoiceNo'].astype(str).str[0].unique()
print("Unique prefixes for negative quantities:", negative_prefixes)


Unique prefixes for positive quantities: ['5']
Unique prefixes for negative quantities: ['C']


# Interpretation of Results

## Positive Quantities:
- **Unique Prefix:** `'5'`  
  This indicates that all positive transactions are associated with invoices starting with `'5'`. This could suggest a specific category of products or a particular sales channel represented by this prefix.

## Negative Quantities:
- **Unique Prefix:** `'C'`  
  The prefix `'C'` is linked to negative quantities, which typically represent returns or discounts. This indicates that these transactions are likely processed under a specific policy or for a particular type of transaction that requires tracking of returns.


In [60]:
cleaned_dataset.shape

(401604, 8)

**Checking For Negative Unit Price**

In [63]:
negative_unitPrice = cleaned_dataset[cleaned_dataset['UnitPrice'] < 0]
print(negative_unitPrice)

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []


> There are no transactions with negative unit prices in your dataset

**Checking For Unique Country Names**

In [67]:
unique_countries = cleaned_dataset['Country'].unique()

# Sorting unique countries
sorted_unique_countries = sorted(unique_countries)

print("Sorted unique country names:")

for countries in sorted_unique_countries:
    print(countries)

Sorted unique country names:
Australia
Austria
Bahrain
Belgium
Brazil
Canada
Channel Islands
Cyprus
Czech Republic
Denmark
EIRE
European Community
Finland
France
Germany
Greece
Iceland
Israel
Italy
Japan
Lebanon
Lithuania
Malta
Netherlands
Norway
Poland
Portugal
RSA
Saudi Arabia
Singapore
Spain
Sweden
Switzerland
USA
United Arab Emirates
United Kingdom
Unspecified


**Generating Additional Columns**

In [70]:
# Total Price
cleaned_dataset.loc[:, 'TotalPrice'] = cleaned_dataset['Quantity'] * cleaned_dataset['UnitPrice']

# Extracting Year and Month from InvoiceDate
cleaned_dataset.loc[:, 'InvoiceDate'] = pd.to_datetime(cleaned_dataset['InvoiceDate'])  # Ensure InvoiceDate is datetime
cleaned_dataset.loc[:, 'Year'] = cleaned_dataset['InvoiceDate'].dt.year
cleaned_dataset.loc[:, 'Month'] = cleaned_dataset['InvoiceDate'].dt.month

# Check for Cancellation
cleaned_dataset.loc[:, 'IsCancelled'] = cleaned_dataset['Quantity'] < 0  # True for cancellations, False otherwise

# Display the updated DataFrame
cleaned_dataset.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  UnitPrice  CustomerID         Country  TotalPrice  \
0 2010-12-01 08:26:00       2.55     17850.0  United Kingdom       15.30   
1 2010-12-01 08:26:00       3.39     17850.0  United Kingdom       20.34   
2 2010-12-01 08:26:00       2.75     17850.0  United Kingdom       22.00   
3 2010-12-01 08:26:00       3.39     17850.0  United Kingdom       20.34   
4 2010-12-01 08:26:00       3.39     17850.0  United Kingdom       20.34   

   Year  Month  IsCancelled  
0  2010     12        False  
1  2010     12        False  
2  2010     12        False  
3  2010     12        False  
4  2010     12        False

## Current DataFrame Overview

Your DataFrame contains the following columns:

- **InvoiceNo**: Unique identifier for each invoice.
- **StockCode**: Unique identifier for each product.
- **Description**: Description of the product.
- **Quantity**: Number of items purchased (can be positive for sales or negative for cancellations).
- **InvoiceDate**: Date and time of the transaction.
- **UnitPrice**: Price per unit of the product.
- **CustomerID**: Unique identifier for each customer.
- **Country**: Country of the customer.
- **TotalPrice**: Total amount for the transaction (calculated as Quantity × UnitPrice).
- **Year**: Year of the transaction (extracted from InvoiceDate).
- **Month**: Month of the transaction (extracted from InvoiceDate).
- **IsCancelled**: Boolean indicating whether the transaction is a cancellation (True if Quantity < 0).


In [73]:
cleaned_dataset.shape

(401604, 12)

In [75]:
cleaned_dataset.to_excel('../data/raw/cleaned_dataset.xlsx', index=False)